### PR TITLE
Home work 2.

### PR DESCRIPTION
**Pull request by interist32**

### CODE REVIEW

File: hw-2/server.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import express = require('express');
+import {UsersStorage} from './storage/users';
```
Comment (MurDaD) [inline hw-2/server.ts]: Surround object properties and array elements with spaces - please add this rule to lint

File: hw-2/server.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      usersStorage.update(user);
+      res.json('Updated');
+    });
+
+app.delete('/users/:id', (req, res) => {
+  usersStorage.remove(req.params['id']);
+  res.json('Softly deleted');
+});
+
+app.get('/getAutoSuggestUsers', (req, res) => {
```
Comment (MurDaD) [inline hw-2/server.ts]: this should be a regular `GET`

Comment (MurDaD) [reply hw-2/server.ts]: I mean, without `getAutoSuggestUsers` addition

File: hw-2/server.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+const PORT = 3000;
+
+const app = express();
+app.use(bodyParser.json());
+
+const validator = createValidator();
+const usersStorage = new UsersStorage();
+
+app.get('/users', (req, res) => {
```
Comment (MurDaD) [inline hw-2/server.ts]: this api can be moved to another file with defining the default prefix (`/users`)

File: hw-2/server.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  res.json(usersStorage.getUsers());
```
Comment (MurDaD) [inline hw-2/server.ts]: please move all the functionality from route file to controller file, to have one separate file with api structure code and another with functionality

File: hw-2/server.ts [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import express = require('express');
+import {UsersStorage} from './storage/users';
+import bodyParser = require('body-parser');
+import {ContainerTypes,
```
Comment (MurDaD) [inline hw-2/server.ts]: `ContainerTypes` should be on the new line - please add lint rule

File: hw-2/server.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        // Use this as a replacement for express.Request
```
Comment (MurDaD) [inline hw-2/server.ts]: there should be no comments in import statement - if you want to comment methods - you should comment them where they are defined

File: hw-2/server.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        ValidatedRequest,
+        // Extend from this to define a valid schema type/interface
+        ValidatedRequestSchema,
+        // Creates a validator that generates middlewares
+        createValidator} from 'express-joi-validation';
```
Comment (MurDaD) [inline hw-2/server.ts]: `} from 'express-joi-validation';` should be on a new line

File: hw-2/server.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+app.get('/users', (req, res) => {
+  res.json(usersStorage.getUsers());
+});
+
+app.get('/users/:id', (req, res) => {
+  res.json(usersStorage.getUserById(req.params.id));
+});
+
+app.post(
+    '/users', validator.body(userSchema),
```
Comment (MurDaD) [inline hw-2/server.ts]: 2 tabs here and following file

**Comment by MurDaD** [comment]
Minor: you can fill memory with dummy data

File: hw-2/server.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  res.json('Softly deleted');
+});
+
+app.get('/getAutoSuggestUsers', (req, res) => {
+  res.json(usersStorage.getUsers(
+      req.query.login, Number(req.query.limit) || undefined));
+});
+
+
+app.listen(PORT, () => {
```
Comment (MurDaD) [inline hw-2/server.ts]: please use `process.env.PORT || PORT`

**Comment by MurDaD** [comment]
User login is not validated - should contain only letters and numbers (that's what I believe, they meant)

**Comment by MurDaD** [comment]
When I create user with the same login, i get `Created` as a response, but should get `400 error`

File: hw-2/server.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+app.put(
+    '/users/:id', validator.body(userSchema),
+    (req: ValidatedRequest<UserRequestSchema>, res) => {
+      const user: User = {
+        ...req.body,
+        id: req.params.id,
+      };
+      usersStorage.update(user);
+      res.json('Updated');
```
Comment (MurDaD) [inline hw-2/server.ts]: it does not return json. ALso good practice to return user object

File: hw-2/server.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+
+app.get('/users/:id', (req, res) => {
+  res.json(usersStorage.getUserById(req.params.id));
+});
+
+app.post(
+    '/users', validator.body(userSchema),
+    (req: ValidatedRequest<UserRequestSchema>, res) => {
+      usersStorage.add(req.body);
+      res.json('Created');
```
Comment (MurDaD) [inline hw-2/server.ts]: it does not return json. ALso good practice to return user id in result

File: hw-2/server.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    });
+
+app.put(
+    '/users/:id', validator.body(userSchema),
+    (req: ValidatedRequest<UserRequestSchema>, res) => {
```
Comment (MurDaD) [inline hw-2/server.ts]: Minor: I'm not able to update only one field, think how to achieve that

Comment (MurDaD) [reply hw-2/server.ts]: Also, I can update `isDeleted` and `id` fields - that should not happen

File: hw-2/storage/users.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    if (login) {
+      users = users.filter(u => u.login.includes(login));
+    }
+    if (limit && limit > 0) {
+      users = users.slice(0, limit);
+    }
+    return users;
+  }
+
+  getUserById(id: string): User|undefined {
```
Comment (MurDaD) [inline hw-2/storage/users.ts]: I can get deleted user - should not happen

**Comment by MurDaD** [comment]
All responses should be json, not string

**Comment by interist32** [comment]
Thank you very much for a review. Updated and pushed again.

**Pull request closed without merging**